server/bleep/src/webserver/studio.rs

```diff
@@ -501,6 +501,61 @@ async fn token_counts(
     })
 }
 
+#[derive(serde::Deserialize)]
+pub struct GetTokenCount {
+    pub path: String,
+    pub repo: RepoRef,
+    pub branch: Option<String>,
+    pub ranges: Option<Vec<Range<usize>>>,
+}
+
+pub async fn get_file_token_count(
+    app: Extension<Application>,
+    Json(params): Json<GetTokenCount>,
+) -> webserver::Result<Json<usize>> {
+    let file = ContextFile {
+        path: params.path,
+        hidden: false,
+        repo: params.repo,
+        branch: params.branch,
+        ranges: params.ranges.unwrap_or_default(),
+    };
+
+    let doc = app
+        .indexes
+        .file
+        .by_path(&file.repo, &file.path, file.branch.as_deref())
+        .await?
+        .with_context(|| {
+            format!(
+                "file `{}` did not exist in repo `{}`, branch `{:?}`",
+                file.path, file.repo, file.branch
+            )
+        })?;
+
+    let mut token_count = 0;
+    let core_bpe = tiktoken_rs::get_bpe_from_model("gpt-4-0613").unwrap();
+
+    if file.ranges.is_empty() {
+        token_count = core_bpe.encode_ordinary(&doc.content).len();
+    } else {
+        let lines = doc.content.lines().collect::<Vec<_>>();
+        for range in &file.ranges {
+            let chunk = lines
+                .iter()
+                .copied()
+                .skip(range.start)
+                .take(range.end - range.start)
+                .collect::<Vec<_>>()
+                .join("\n");
+
+            token_count += core_bpe.encode_ordinary(&chunk).len();
+        }
+    }
```

Comment:
Can we extract this into a function? We can use the function [above](https://github.com/BloopAI/bloop/pull/935/files#diff-663c689783c1e5c15f880b3602f7455c2dbc19b71245624756e4d8fe11fcb689R446-R466) as well, to avoid duplication of this token counting logic.

Reply:
done